feat: add udc to options

CodeChenL opened the pull request:
关联 #11

RadxaYuntian commented on debian/radxa-otgutils.radxa-adbd@.service:
https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#PropagatesReloadTo=

应该需要传递reload和stop事件吧

RadxaYuntian commented on debian/radxa-otgutils.radxa-udc@.service:
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=Enable UDC on %i
+Documentation=https://github.com/radxa-pkg/radxa-otgutils/
+After=network.target radxa-adbd@%i.service radxa-usbnet@%i.service
```
```suggestion
After=network.target radxa-adbd@%i.service radxa-ncm@%i.service radxa-ecm@%i.service
```

RadxaYuntian commented on debian/radxa-otgutils.radxa-ecm@.service:
```diff
@@ -0,0 +1,16 @@
+
+[Install]
+WantedBy=multi-user.target
+Also=radxa-udc@%i.service
```
```suggestion
Alias=radxa-usbnet@%i.service
Also=radxa-udc@%i.service
```

RadxaYuntian commented on src/radxa-otgutils:
```diff
@@ -112,19 +117,16 @@ function_stop()
 {
     lock_udc
-    configfs_init
-    function_init
```
这个没必要移动位置吧，正常开启多个服务后本身就会多次执行这两个命令，在udc start时再执行一次应该不影响

RadxaYuntian commented on src/radxa-otgutils:
```diff
@@ -134,35 +136,46 @@ start()
 stop()
 {
     lock_udc
-    function_stop
```
跟上面同理

RadxaYuntian reviewed (commented):
看上去没什么问题，有测试实际使用的效果么？
另外也要验证如果之前开启了radxa-usbnet服务，升级后能否保持开启

CodeChenL commented:
> 看上去没什么问题，有测试实际使用的效果么？ 另外也要验证如果之前开启了radxa-usbnet服务，升级后能否保持开启

需要加个postinst来重新enable otg服务

RadxaYuntian commented:
> > 看上去没什么问题，有测试实际使用的效果么？ 另外也要验证如果之前开启了radxa-usbnet服务，升级后能否保持开启
> 
> 需要加个postinst来重新enable otg服务

要在这里安装：https://github.com/radxa-pkg/radxa-otgutils/blob/main/debian/rules#L13-L14

CodeChenL commented:
> > > 看上去没什么问题，有测试实际使用的效果么？ 另外也要验证如果之前开启了radxa-usbnet服务，升级后能否保持开启
> > 
> > 
> > 需要加个postinst来重新enable otg服务
> 
> 要在这里安装：https://github.com/radxa-pkg/radxa-otgutils/blob/main/debian/rules#L13-L14

试了没效果

CodeChenL commented:
> > > > 看上去没什么问题，有测试实际使用的效果么？ 另外也要验证如果之前开启了radxa-usbnet服务，升级后能否保持开启
> > > 
> > > 
> > > 需要加个postinst来重新enable otg服务
> > 
> > 
> > 要在这里安装：https://github.com/radxa-pkg/radxa-otgutils/blob/main/debian/rules#L13-L14
> 
> 试了没效果

失效的软链接
```bash
ock@rock-2a:/$ ls -l /etc/systemd/system/multi-user.target.wants/radxa-usbnet@fe500000.dwc3.service
lrwxrwxrwx 1 root root 41 Jun 21 08:26 /etc/systemd/system/multi-user.target.wants/radxa-usbnet@fe500000.dwc3.service -> /lib/systemd/system/radxa-usbnet@.service
rock@rock-2a:/$
```

CodeChenL commented:
只能enable ecm 然后才能 restart usbnet，手动也enbale不了（忽略e20c的串口回显错误）
```bash
rock@rock-2a:/$ sdo systemctl enable radxa-usbnet@fe500000.dwc3
Failed to enable unit: Refusing to operate on alias name or linked unit file: radxa-usbnet@fe500000.dwc3.service
rock@rock-2a:/$
```

RadxaYuntian commented:
这是两个问题：
1. 需要将systemd unit在rules里面添加，否则不会被打包
2. 可能要分析一下为什么bind9就可以重命名他们的服务：https://salsa.debian.org/dns-team/bind9/-/commit/6fd962a36bb54e0106e001aff6cce056d54e2526#befe0c97e7c2864a71511033f5078fad6c97cfc1

有可能只能把usbnet恢复，然后加一个ecm的alias

CodeChenL commented:
> 这是两个问题：
> 
> 1. 需要将systemd unit在rules里面添加，否则不会被打包
> 2. 可能要分析一下为什么bind9就可以重命名他们的服务：https://salsa.debian.org/dns-team/bind9/-/commit/6fd962a36bb54e0106e001aff6cce056d54e2526#befe0c97e7c2864a71511033f5078fad6c97cfc1
> 
> 有可能只能把usbnet恢复，然后加一个ecm的alias

1. 尝试过在rule install usbnet但是没效果，实际的软件包里没有

RadxaYuntian commented:
你现在没有usbnet这个文件所以不行啊

CodeChenL commented:
> 你现在没有usbnet这个文件所以不行啊

替换之后,升级前先enable usbnet，升级之后udc不会自动enbale

CodeChenL commented:
可能还是得加个postinst？

RadxaYuntian commented:
Requires= & Before=

RadxaYuntian commented:
Also=可能是不需要的

RadxaYuntian commented:
现在可以更新后正常启动cdc服务么？

CodeChenL commented:
> 现在可以更新后正常启动cdc服务么？

可以

CodeChenL commented:
这是先开启adbd与usbnet，更新以后的各个服务状态：
```bash
rock@rock-5b:~$ sudo systemctl status radxa-*
● radxa-adbd@fc000000.usb.service - Enable adbd on fc000000.usb
     Loaded: loaded (/lib/systemd/system/radxa-adbd@.service; enabled; preset: enabled)
     Active: active (exited) since Wed 2024-06-26 06:40:00 UTC; 9min ago
       Docs: https://github.com/radxa-pkg/radxa-otgutils/
    Process: 1296 ExecStart=/usr/sbin/radxa-otgutils adbd fc000000.usb start (code=exited, status=0/SUCCESS)
   Main PID: 1296 (code=exited, status=0/SUCCESS)
      Tasks: 4 (limit: 9193)
     Memory: 2.5M
        CPU: 47ms
     CGroup: /system.slice/system-radxa\x2dadbd.slice/radxa-adbd@fc000000.usb.service
             └─1334 /usr/bin/adbd

Jun 26 06:40:00 rock-5b systemd[1]: Started radxa-adbd@fc000000.usb.service - Enable adbd on fc000000.usb.
Jun 26 06:40:00 rock-5b radxa-otgutils[1296]: /sys/kernel/config/usb_gadget/fc000000.usb /
Jun 26 06:40:00 rock-5b radxa-otgutils[1296]: /

● radxa-ecm@fc000000.usb.service - Enable USB CDC Ethernet Control Model on fc000000.usb
     Loaded: loaded (/lib/systemd/system/radxa-ecm@.service; disabled; preset: enabled)
     Active: active (exited) since Wed 2024-06-26 06:40:00 UTC; 9min ago
       Docs: https://github.com/radxa-pkg/radxa-otgutils/
    Process: 1298 ExecStart=/usr/sbin/radxa-otgutils ecm fc000000.usb start (code=exited, status=0/SUCCESS)
   Main PID: 1298 (code=exited, status=0/SUCCESS)
        CPU: 28ms

Jun 26 06:40:00 rock-5b systemd[1]: Started radxa-ecm@fc000000.usb.service - Enable USB CDC Ethernet Control Model on fc000000.usb.
Jun 26 06:40:00 rock-5b radxa-otgutils[1298]: /sys/kernel/config/usb_gadget/fc000000.usb /
Jun 26 06:40:00 rock-5b radxa-otgutils[1298]: /

● radxa-udc@fc000000.usb.service - Enable UDC on fc000000.usb
     Loaded: loaded (/lib/systemd/system/radxa-udc@.service; disabled; preset: enabled)
     Active: active (exited) since Wed 2024-06-26 06:40:00 UTC; 9min ago
       Docs: https://github.com/radxa-pkg/radxa-otgutils/
    Process: 1299 ExecStart=/usr/sbin/radxa-otgutils udc fc000000.usb start (code=exited, status=0/SUCCESS)
   Main PID: 1299 (code=exited, status=0/SUCCESS)
        CPU: 29ms

Jun 26 06:40:00 rock-5b systemd[1]: Started radxa-udc@fc000000.usb.service - Enable UDC on fc000000.usb.
Jun 26 06:40:00 rock-5b radxa-otgutils[1299]: /sys/kernel/config/usb_gadget/fc000000.usb /
Jun 26 06:40:00 rock-5b radxa-otgutils[1299]: /

● radxa-usbnet@fc000000.usb.service - Enable USB CDC Ethernet Control Model on fc000000.usb
     Loaded: loaded (/lib/systemd/system/radxa-usbnet@.service; enabled; preset: enabled)
     Active: active (exited) since Wed 2024-06-26 06:40:00 UTC; 9min ago
       Docs: https://github.com/radxa-pkg/radxa-otgutils/
    Process: 1297 ExecStart=/usr/bin/echo radxa-usbnet@fc000000.usb.service will be discarded, please use radxa-ecm@fc000000.usb.service instead (code=exited, status=0/SUCCESS)
   Main PID: 1297 (code=exited, status=0/SUCCESS)
        CPU: 2ms

Jun 26 06:40:00 rock-5b systemd[1]: Started radxa-usbnet@fc000000.usb.service - Enable USB CDC Ethernet Control Model on fc000000.usb.
Jun 26 06:40:00 rock-5b echo[1297]: radxa-usbnet@fc000000.usb.service will be discarded, please use radxa-ecm@fc000000.usb.service instead
rock@rock-5b:~$ 
```

RadxaYuntian commented on debian/radxa-otgutils.radxa-ecm@.service:
```diff
@@ -0,0 +1,20 @@
+
+[Install]
+WantedBy=multi-user.target
+Alias=radxa-usbnet@%i.service
```
有两个服务那就没法alias了，不然我enable radxa-usbnet到底会开启哪一个？

RadxaYuntian commented on debian/radxa-otgutils.radxa-usbnet@.service:
https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html

> Units can be aliased (have an alternative name), by creating a symlink from the new name to the existing name in one of the unit search paths. For example, systemd-networkd.service has the alias dbus-org.freedesktop.network1.service, created during installation as a symlink, so when systemd is asked through D-Bus to load dbus-org.freedesktop.network1.service, it'll load systemd-networkd.service. As another example, default.target — the default system target started at boot — is commonly aliased to either multi-user.target or graphical.target to select what is started by default. Alias names may be used in commands like disable, start, stop, status, and similar, and in all unit dependency directives, including Wants=, Requires=, Before=, After=. Aliases cannot be used with the preset command.

看看是不是用symlink代替了

https://manpages.debian.org/testing/debhelper/dh_link.1.en.html

RadxaYuntian commented on debian/rules:
```diff
@@ -11,4 +11,8 @@ override_dh_builddeb:
+	dh_installsystemd --name=radxa-ncm@
+	dh_installsystemd --name=radxa-udc@
 	dh_installsystemd --name=radxa-usbnet@
+	dh_installsystemd --remaining
```
这个到底有啥用？是不是没有必要

RadxaYuntian commented on debian/rules:
```diff
@@ -14,5 +14,4 @@ override_dh_installsystemd:
 	dh_installsystemd --name=radxa-udc@
-	dh_installsystemd --name=radxa-usbnet@
-	dh_installsystemd --remaining
+	dh_link usr/lib/systemd/system/radxa-ecm@.service usr/lib/systemd/system/radxa-usbnet@.service
```
dh_link有对应的配置文件，写在Files里面了
但感觉Install.Alias就应该会创建symlink啊：
> Unit files may specify aliases through the Alias= directive in the [Install] section. When the unit is enabled, symlinks will be created for those names, and removed when the unit is disabled. For example, reboot.target specifies Alias=ctrl-alt-del.target, so when enabled, the symlink /etc/systemd/system/ctrl-alt-del.service pointing to the reboot.target file will be created, and when Ctrl+Alt+Del is invoked, systemd will look for the ctrl-alt-del.service and execute reboot.service. systemd does not look at the [Install] section at all during normal operation, so any directives in that section only have an effect through the symlinks created during enablement.

CodeChenL replied:
之前有试过，没效果
> 但感觉Install.Alias就应该会创建symlink啊：

RadxaYuntian replied:
哦好像只限于enable那一刻：When the unit is enabled, symlinks will be created for those names

RadxaYuntian commented:
我现在在处理更新后服务迁移的问题，但服务本身有一个不受支持的属性：
```
 /lib/systemd/system/radxa-ecm@.service:6: Unknown key name 'PropagatesStopTo' >
```
可能要测试一下stop,然后看看这个怎么处理

CodeChenL commented:
> 我现在在处理更新后服务迁移的问题，但服务本身有一个不受支持的属性：
> 
> ```
>  /lib/systemd/system/radxa-ecm@.service:6: Unknown key name 'PropagatesStopTo' >
> ```
> 
> 可能要测试一下stop,然后看看这个怎么处理


应该是bullseye的systemd的版本247太低了https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#PropagatesStopTo=

CodeChenL commented:
> > 我现在在处理更新后服务迁移的问题，但服务本身有一个不受支持的属性：
> > ```
> >  /lib/systemd/system/radxa-ecm@.service:6: Unknown key name 'PropagatesStopTo' >
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 可能要测试一下stop,然后看看这个怎么处理
> 
> 应该是bullseye的systemd的版本247太低了https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#PropagatesStopTo=

我在bookworm测试可用

RadxaYuntian commented:
那先不处理这个问题

RadxaYuntian commented:
添加了升级迁移的脚本。在debian 12上再试试

CodeChenL commented:
> 添加了升级迁移的脚本。在debian 12上再试试

可正常升级
```bash
rock@rock-5b:/$ cat ~/log 
● radxa-ecm@fc000000.usb.service - Enable USB CDC Ethernet Control Model on fc000000.usb
     Loaded: loaded (/lib/systemd/system/radxa-ecm@.service; disabled; preset: enabled)
     Active: active (exited) since Thu 2024-07-11 02:54:37 UTC; 2min 5s ago
       Docs: https://github.com/radxa-pkg/radxa-otgutils/
    Process: 746 ExecStart=/usr/sbin/radxa-otgutils ecm fc000000.usb start (code=exited, status=0/SUCCESS)
   Main PID: 746 (code=exited, status=0/SUCCESS)
        CPU: 28ms

Jul 11 02:54:37 rock-5b systemd[1]: Started radxa-ecm@fc000000.usb.service - Enable USB CDC Ethernet Control Model on fc000000.usb.
Jul 11 02:54:37 rock-5b radxa-otgutils[746]: /sys/kernel/config/usb_gadget/fc000000.usb /
Jul 11 02:54:37 rock-5b radxa-otgutils[746]: /

● radxa-adbd@fc000000.usb.service - Enable adbd on fc000000.usb
     Loaded: loaded (/lib/systemd/system/radxa-adbd@.service; enabled; preset: enabled)
     Active: active (exited) since Thu 2024-07-11 02:54:37 UTC; 2min 5s ago
       Docs: https://github.com/radxa-pkg/radxa-otgutils/
    Process: 745 ExecStart=/usr/sbin/radxa-otgutils adbd fc000000.usb start (code=exited, status=0/SUCCESS)
   Main PID: 745 (code=exited, status=0/SUCCESS)
      Tasks: 4 (limit: 9192)
     Memory: 2.4M
        CPU: 57ms
     CGroup: /system.slice/system-radxa\x2dadbd.slice/radxa-adbd@fc000000.usb.service
             └─782 /usr/bin/adbd

Jul 11 02:54:37 rock-5b systemd[1]: Started radxa-adbd@fc000000.usb.service - Enable adbd on fc000000.usb.
Jul 11 02:54:37 rock-5b radxa-otgutils[745]: /sys/kernel/config/usb_gadget/fc000000.usb /
Jul 11 02:54:37 rock-5b radxa-otgutils[745]: /

● radxa-udc@fc000000.usb.service - Enable UDC on fc000000.usb
     Loaded: loaded (/lib/systemd/system/radxa-udc@.service; disabled; preset: enabled)
     Active: active (exited) since Thu 2024-07-11 02:54:37 UTC; 2min 5s ago
       Docs: https://github.com/radxa-pkg/radxa-otgutils/
    Process: 750 ExecStart=/usr/sbin/radxa-otgutils udc fc000000.usb start (code=exited, status=0/SUCCESS)
   Main PID: 750 (code=exited, status=0/SUCCESS)
        CPU: 31ms

Jul 11 02:54:37 rock-5b systemd[1]: Started radxa-udc@fc000000.usb.service - Enable UDC on fc000000.usb.
Jul 11 02:54:37 rock-5b radxa-otgutils[750]: /sys/kernel/config/usb_gadget/fc000000.usb /
Jul 11 02:54:37 rock-5b radxa-otgutils[750]: /
```

RadxaYuntian commented:
测试一下之前说的otg无法开启的问题是否还会复现，不会的话重写一下历史

CodeChenL commented:
> 测试一下之前说的otg无法开启的问题是否还会复现，不会的话重写一下历史

其实这个问题，我之前添加失败重启就解决了 https://github.com/radxa-pkg/radxa-otgutils/commit/9dd8c0a16fbef97015a2f5cce5b0136d1df6f291

RadxaYuntian commented:
那整理一下提交吧

RadxaYuntian commented:
另外测试一下能不能在rsetup里面正常开关功能
debian 11 & 12

CodeChenL commented:
> rsetup



> 另外测试一下能不能在rsetup里面正常开关功能 debian 11 & 12

rsetup还得改这行https://github.com/radxa-pkg/rsetup/blob/09b336f68522c77019a3df187d8678bfc2fd8da3/src/usr/lib/rsetup/tui/hardware/hardware.sh#L393

RadxaYuntian commented:
主要是看systemd sevice能不能正常传递重启信号，和rsetup自身关系不太大

CodeChenL commented:
可以使用rsetup正常开关

RadxaYuntian commented:
11 等下测试一下

CodeChenL commented:
> 11 等下测试一下

debian 11正常

RadxaYuntian commented:
1. 检查一下为什么没有清除老的radxa-usbnet.service
2. 看看如何在debian 11上stop start后能正常开启服务而不用手动restart udc服务

CodeChenL commented:
> 1. 检查一下为什么没有清除老的radxa-usbnet.service
> 2. 看看如何在debian 11上stop start后能正常开启服务而不用手动restart udc服务

可以使用https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#ExecStartPre= 来替换https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#PropagatesStopTo= 来提高兼容性

CodeChenL commented:
> 1. 检查一下为什么没有清除老的radxa-usbnet.service
> 2. 看看如何在debian 11上stop start后能正常开启服务而不用手动restart udc服务

目前能做到第二点来，但是同时开启两个funciton时，暂停一个funciton，udc也会跟着暂停，导致另一个funciton用不了

CodeChenL commented:
> > 1. 检查一下为什么没有清除老的radxa-usbnet.service
> > 2. 看看如何在debian 11上stop start后能正常开启服务而不用手动restart udc服务
> 
> 目前能做到第二点来，但是同时开启两个funciton时，暂停一个funciton，udc也会跟着暂停

RadxaYuntian commented on src/radxa-otgutils:
```diff
@@ -90,6 +95,9 @@ function_stop()
             local GADGET="$MODE.usb0"
             ;;
+        *)
+            return 0
```
这个函数怎么会被传入这种不支持的值？

CodeChenL replied:
会传入udc

RadxaYuntian replied:
那应该明确写明是udc

RadxaYuntian commented on src/radxa-otgutils:
```diff
@@ -15,6 +15,11 @@ unlock_udc()
+    if [ -z "$(find "/sys/kernel/config/usb_gadget/$UDC/functions" -mindepth 1 -maxdepth 1)" ];
+    then
+        echo "No functions defined, ignore enable UDC"
+        return 0
```
这里没必要报错吧，就开一个没有功能的udc呗

CodeChenL replied:
没有功能的udc内核会报错

RadxaYuntian commented on debian/radxa-otgutils.radxa-udc@.service:
```diff
@@ -1,6 +1,7 @@
 [Unit]
 Description=Enable UDC on %i
 Documentation=https://github.com/radxa-pkg/radxa-otgutils/
+PartOf=radxa-adbd@%i.service radxa-ncm@%i.service radxa-ecm@%i.service
```
应该是这个导致一个服务关闭后udc整个都关闭了，所以这里不能这么写
debian 11支持不是必须的，我们可以把这个版本仅限于debian 12上使用

CodeChenL replied:
debian12删掉这段也是这种情况

CodeChenL replied:
PartOf的效果等同debian12的PropagatesStopTo+PropagatesRestartTo（这个属性是虚构的）

RadxaYuntian replied:
PartOf在udc是明显错误的，这导致udc依赖adb而不是反过来。应该还是在adb里面Requires udc
如何在adb关闭后重启udc我还要再看看

RadxaYuntian reviewed (changes requested):
另外看看为什么老的服务没有被uninstall掉

CodeChenL commented:
> 另外看看为什么老的服务没有被uninstall掉

老服务可以正常uninstall

RadxaYuntian commented:
回头再一起测试一下，另外可以考虑使用StopWhenUnneeded=

CodeChenL commented:
如果ExecStartPost能正常用就好了，但是我用的时候启动超时了才会执行

RadxaYuntian commented:
尽量避免自己处理服务信号吧，让systemd统一调配，不然可能会在启动时反复启动udc服务多次

CodeChenL commented:
> StopWhenUnneeded

得删掉所有依赖关系才能用

RadxaYuntian commented:
https://superuser.com/questions/849001/why-does-systemd-process-stop-as-soon-as-it-starts-for-foreman-server-process

看看是不是要加一个udc.target？

RadxaYuntian commented:
> So the problem is with StopWhenUnneeded= in the main app.target unit.  Since you start it **manually**, it never has any other units depending on it, therefore it will be stopped immediately.

不能手动启动udc服务

RadxaYuntian commented on debian/radxa-otgutils.radxa-adbd@.service:
```diff
@@ -2,12 +2,16 @@
 Description=Enable adbd on %i
 Documentation=https://github.com/radxa-pkg/radxa-otgutils/
 After=network.target
```
这几个服务应该都不需要after,在udc那边after就好了

RadxaYuntian commented on debian/radxa-otgutils.radxa-udc@.service:
```diff
@@ -0,0 +1,17 @@
+Description=Enable UDC on %i
+Documentation=https://github.com/radxa-pkg/radxa-otgutils/
+PartOf=radxa-adbd@%i.service radxa-ncm@%i.service radxa-ecm@%i.service
+After=network.target radxa-adbd@%i.service radxa-ncm@%i.service radxa-ecm@%i.service
```
这里应该只用定义udc本身的after就好，其他的服务自己定义了before,不用重复定义

RadxaYuntian commented on debian/radxa-otgutils.radxa-adbd@.service:
```diff
@@ -1,11 +1,9 @@
 PropagatesReloadTo=radxa-udc@%i.service
 PropagatesStopTo=radxa-udc@%i.service
 Wants=radxa-udc@%i.service
-Before=radxa-udc@%i.service
```
这里的before留着啊，新加服务直接复制一下就好。现在的话还要去改cdc的after

RadxaYuntian commented:
可能要用ExecStopPost重启一下udc服务，并且删掉PropagatesStopTo和PropagatesReloadTo

RadxaYuntian commented:
以及ExecStartPost

CodeChenL commented:
> 可能要用ExecStopPost重启一下udc服务，并且删掉PropagatesStopTo和PropagatesReloadTo

试过了，会等到stop超时才会执行，除非去掉依赖关系

RadxaYuntian commented:
> > 可能要用ExecStopPost重启一下udc服务，并且删掉PropagatesStopTo和PropagatesReloadTo
> 
> 试过了，会等到stop超时才会执行，除非去掉依赖关系

我看[这里](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#ExecStopPost=)的说明应该是stop结束后就会运行。你看看之前的问题是不是跟type不是oneshot相关

CodeChenL commented:
> > > 可能要用ExecStopPost重启一下udc服务，并且删掉PropagatesStopTo和PropagatesReloadTo
> > 
> > 
> > 试过了，会等到stop超时才会执行，除非去掉依赖关系
> 
> 我看[这里](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#ExecStopPost=)的说明应该是stop结束后就会运行。你看看之前的问题是不是跟type不是oneshot相关

type也之前试过换了好几种也没效果

CodeChenL commented:
只能运行/usr/sbin/radxa-otgutils udc %i restart，不能运行/usr/bin/systemctl restart radxa-udc@%i.service

CodeChenL commented:
运行的 post start systemctl的状态：
```
○ radxa-udc@fc000000.usb.service - Enable UDC on fc000000.usb
     Loaded: loaded (/lib/systemd/system/radxa-udc@.service; disabled; preset: enabled)
     Active: inactive (dead)
       Docs: https://github.com/radxa-pkg/radxa-otgutils/

● radxa-adbd@fc000000.usb.service - Enable adbd on fc000000.usb
     Loaded: loaded (/lib/systemd/system/radxa-adbd@.service; enabled; preset: enabled)
     Active: activating (start-post) since Thu 2024-07-18 11:01:16 UTC; 14s ago
       Docs: https://github.com/radxa-pkg/radxa-otgutils/
    Process: 912 ExecStart=/usr/sbin/radxa-otgutils adbd fc000000.usb start (code=exited, st>
   Main PID: 912 (code=exited, status=0/SUCCESS); Control PID: 1153 (systemctl)
      Tasks: 5 (limit: 9193)
     Memory: 5.0M
        CPU: 72ms
     CGroup: /system.slice/system-radxa\x2dadbd.slice/radxa-adbd@fc000000.usb.service
             ├─1033 /usr/bin/adbd
             └─1153 /usr/bin/systemctl restart radxa-udc@fc000000.usb.service

Jul 18 11:01:16 rock-5b systemd[1]: Starting radxa-adbd@fc000000.usb.service - Enable adbd o>
Jul 18 11:01:16 rock-5b radxa-otgutils[912]: /sys/kernel/config/usb_gadget/fc000000.usb /
Jul 18 11:01:16 rock-5b radxa-otgutils[912]: /

```

RadxaYuntian commented:
会不会是里面文件夹flock导致的死锁？加上打印试试

CodeChenL commented:
> 会不会是里面文件夹flock导致的死锁？加上打印试试

```bash
rock@rock-5b:/$ sudo systemctl status radxa-*|cat
● radxa-adbd@fc000000.usb.service - Enable adbd on fc000000.usb
     Loaded: loaded (/lib/systemd/system/radxa-adbd@.service; enabled; preset: enabled)
    Drop-In: /etc/systemd/system/radxa-adbd@.service.d
             └─override.conf
     Active: activating (start-post) since Fri 2024-07-19 02:51:45 UTC; 1min 48s ago
       Docs: https://github.com/radxa-pkg/radxa-otgutils/
    Process: 1815 ExecStart=/usr/sbin/radxa-otgutils adbd fc000000.usb start (code=exited, status=0/SUCCESS)
   Main PID: 1815 (code=exited, status=0/SUCCESS); Control PID: 1833 (systemctl)
      Tasks: 5 (limit: 9193)
     Memory: 3.3M
        CPU: 81ms
     CGroup: /system.slice/system-radxa\x2dadbd.slice/radxa-adbd@fc000000.usb.service
             ├─1828 /usr/bin/adbd
             └─1833 /usr/bin/systemctl reload radxa-udc@fc000000.usb.service

Jul 19 02:51:45 rock-5b systemd[1]: Starting radxa-adbd@fc000000.usb.service - Enable adbd on fc000000.usb...
Jul 19 02:51:45 rock-5b radxa-otgutils[1815]: Locking UDC
Jul 19 02:51:45 rock-5b radxa-otgutils[1815]: Lock UDC
Jul 19 02:51:45 rock-5b radxa-otgutils[1815]: /sys/kernel/config/usb_gadget/fc000000.usb /
Jul 19 02:51:45 rock-5b radxa-otgutils[1815]: /
Jul 19 02:51:46 rock-5b radxa-otgutils[1815]: Unlocking UDC
Jul 19 02:51:46 rock-5b radxa-otgutils[1815]: Unlock UDC

○ radxa-udc@fc000000.usb.service - Enable UDC on fc000000.usb
     Loaded: loaded (/lib/systemd/system/radxa-udc@.service; disabled; preset: enabled)
     Active: inactive (dead)
       Docs: https://github.com/radxa-pkg/radxa-otgutils/
rock@rock-5b:/$ 
```

RadxaYuntian commented:
另外systemd那几个额外参数现在有结论了么？到底支持情况是如何的

CodeChenL commented:
> 另外systemd那几个额外参数现在有结论了么？到底支持情况是如何的

虽然在debian12上会报错，但是应该是警告级别的，不会打断处理脚本，执行效果也是一样的

RadxaYuntian commented on debian/radxa-otgutils.radxa-adbd@.service:
```diff
@@ -2,7 +2,7 @@
 Documentation=https://github.com/radxa-pkg/radxa-otgutils/
 Before=radxa-udc@%i.service
-Requires=radxa-udc@%i.service
+Wants=radxa-udc@%i.service
```
这里为什么又改了？

CodeChenL replied:
因为startpost restart udc, 会导致funciton也跟着重启，然后死循环了

RadxaYuntian commented on debian/radxa-otgutils.radxa-adbd@.service:
```diff
@@ -10,6 +10,7 @@ RemainAfterExit=yes
 ExecStart=/usr/sbin/radxa-otgutils adbd %i start
 ExecStartPost=/usr/bin/bash -c "systemctl try-restart radxa-udc@%i.service &"
 ExecStop=/usr/sbin/radxa-otgutils adbd %i stop
+ExecStopPost=/usr/bin/bash -c "systemctl try-restart radxa-udc@%i.service &"
```
自己测试一下吧，至少ecm不需要重启udc服务usb0就没了

CodeChenL replied:
我这里测试正常的

RadxaYuntian reviewed (changes requested):
感觉现在就是没有办法把udc的功能独立出去，我这边再改一下吧

RadxaYuntian reviewed (commented):
把UDC配置的功能又加回每个function里面了，另外把之前的UDC服务改成了一个target来管理和其他服务的依赖关系。麻烦再次测试一下安装时会不会清理旧的服务链接，以及服务现在启动停止功能是否正常

RadxaYuntian commented on src/radxa-otgutils:
```diff
@@ -148,9 +149,11 @@ function_stop()
 start()
 {
     lock_udc
+    disable_udc
```
哪里需要先禁用了？另外这里修改是不是stop也要对应修改？

CodeChenL replied:
添加funciton需要先stop，减少funciton不需要

RadxaYuntian commented on debian/radxa-otgutils.radxa-adbd@.service:
```diff
@@ -3,6 +3,7 @@ Description=Enable adbd on %i
 Documentation=https://github.com/radxa-pkg/radxa-otgutils/
 
 [Service]
+RemainAfterExit=yes
```
没有这个是不是不能systemctl stop了？

RadxaYuntian replied:
我自己测试好像没问题

CodeChenL replied:
我这里不加这个会运行完会自动运行stop

RadxaYuntian replied:
那还是要用
```
Type=oneshot
RemainAfterExit=yes
```

RadxaYuntian reviewed (changes requested):
整理一下提交吧，应该会分成几个

CodeChenL committed:
feat: create the reuse function disable_udc

CodeChenL committed:
fix: udc is not enabled when funcitons is empty

CodeChenL committed:
refactor: merge configfs_init to configfs_make

CodeChenL committed:
style: optimize bash syntax

CodeChenL committed:
fix: optimize UDC on/off logic

CodeChenL committed:
feat: rename usbnet to ecm and add ncm support

CodeChenL committed:
fix: optimize systemd service files

1. rename radxa-usbnet to radxa-ecm
2. delete、relink old service files
3. add radxa-otgutils.radxa-otgutils.target
